jscomp/others/belt_String.mli

```diff
@@ -0,0 +1,326 @@
+]}
+*)
+external replaceRegex : t -> old:Js_re.t -> by:t -> t = "replace" [@@bs.send]
+
+
+(**
+  [matchRegex str regexp] matches a string against the given [regexp]. If there is no match, it returns [None].
+  For regular expressions without the [g] modifier, if there is a match, the return value is [Some array] where the array contains:
+  
+  {ul
+    {- The entire matched string}
+    {- Any capture groups if the [regexp] had parentheses}
+  }
+  
+  For regular expressions with the [g] modifier, a matched expression returns [Some array] with all the matched substrings and no capture groups.
+  
+@example {[
+  matchRegex "The better bats" [%re "/b[aeiou]t/"] = Some [|"bet"|]
+  matchRegex "The better bats" [%re "/b[aeiou]t/g"] = Some [|"bet";"bat"|]
+  matchRegex "Today is 2018-04-05." [%re "/(\\d+)-(\\d+)-(\\d+)/"] =
+    Some [|"2018-04-05"; "2018"; "04"; "05"|]
+  matchRegex "The large container." [%re "/b[aeiou]g/"] = None
+]}
+
+*)
+external matchRegex : string -> Js_re.t -> t array option = "match" [@@bs.send] [@@bs.return {null_to_opt}]
+
+(**
+  [split delimiter str] splits the given [str] at every occurrence of [delimiter] and returns an
+  array of the resulting substrings.
+  
+@example {[
+  split "2018-01-02" "-" = [|"2018"; "01"; "02"|];;
+  split "a,b,,c" "," = [|"a"; "b"; ""; "c"|];;
+  split "good::bad as great::awful" "::" = [|"good"; "bad as great"; "awful"|];;
+  split "has-no-delimiter" ";"  = [|"has-no-delimiter"|];;
+]};
+*)
+external split : t ->  t -> t array  = "split" [@@bs.send]
+
+(**
+  [splitAtMost str delimiter n] splits the given [str] at every occurrence of [delimiter] and 
+  returns an array of the first [n] resulting substrings. If [n] is negative or greater than the
+   number of substrings, the array will contain all the substrings.
+  
+@example {[
+  splitAtMost "ant/bee/cat/dog/elk" "/" 3  = [|"ant"; "bee"; "cat"|];;
+  splitAtMost "ant/bee/cat/dog/elk" "/"  0  = [| |];;
+  splitAtMost "ant/bee/cat/dog/elk" "/" 9  = [|"ant"; "bee"; "cat"; "dog"; "elk"|];;
+]}
+*)
+external splitAtMost: t -> t -> int -> t array = "split" [@@bs.send]
+
+
+(** ES2015:
+    [startsWith str substr] returns [true] if the [str] starts with [substr], [false] otherwise.
+    
+@example {[
+  startsWith "BuckleScript" "Buckle"  = true;;
+  startsWith "BuckleScript" ""  = true;;
+  startsWith "JavaScript" "Buckle"  = false;;
+]} 
+*)
+external startsWith : t -> t -> bool = "startsWith" [@@bs.send]
+
+
+(**
+  [substr str ~from: n] returns the substring of [str] from position [n] to the end of the string.
+  
+  If [n] is less than zero, the starting position is the length of [str] - [n].
+  
+  If [n] is greater than or equal to the length of [str], returns the empty string.
+  
+@example {[
+  substr "abcdefghij" ~from: 3 ~len: 2 = "de"
+  substr "abcdefghij" ~from: (-3) ~len: 3 = "hij"
+  substr "abcdefghij" ~from: 12 ~len: 2= ""
+]}
+*)
+external substr : t -> from:int -> len:int -> t = "substr" [@@bs.send]
```

Comment:
given `String.prototype.substring` is equivalent to `String.prototype.slice` (`index` + `endIndex`) and `String.prototype.substr` is `index` + `length`, I figured it would be interesting to leave the two ways of doing so. 

if the `substr` signature isn't really interesting for users (I'm not sure if widely used), we can remove it altogether.

Reply:
Yeah considering the fact that substring≃slice, feel free to forget what I said :)

Nice read about diff between substring & slice https://stackoverflow.com/questions/2243824/what-is-the-difference-between-string-slice-and-string-substring#2243835